src/wp-includes/global-styles-and-settings.php

```diff
@@ -264,6 +264,26 @@ function ( $item ) {
  * @return bool Returns true if theme or its parent has a theme.json file, false otherwise.
  */
 function wp_theme_has_theme_json() {
+	static $theme_has_support = null;
+
+	if (
+		null !== $theme_has_support &&
+		/*
+		 * Ignore static cache when `WP_DEBUG` is enabled. Why? To avoid interfering with
+		 * the theme developer's workflow.
+		 *
+		 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+		 */
+		! WP_DEBUG &&
+		/*
+		 * Ignore cache when automated test suites are running. Why? To ensure
+		 * the static cache is reset between each test.
+		 */
+		! ( defined( 'WP_RUN_CORE_TESTS' ) && WP_RUN_CORE_TESTS )
```

Comment:
Instead of hardcode this here, why not just add a filter?

```
apply_filters( 'wp_theme_use_persistent_cache', true )
```

In the the phpunit test bootstrap, just add. 

```
add_filters( 'wp_theme_use_persistent_cache', '__return_false' );
```

Reply:
Example 
https://github.com/WordPress/wordpress-develop/blob/d19f5f01701838f4def68f94b580461c62f4babd/tests/phpunit/includes/bootstrap.php#L282-L289

Reply:
@spacedmonkey Why not add the filter?

* The constants encapsulate the conditions within the function internals for ease of iteration later if needed.
* BC concerns for the filter.
* Performance: the overhead of invoking a filter each time the function is called.
* At this time, it's unknown if a publicly available reset is needed within Core and/or for theme/plugin developers.

There is precedence for using the constants within Core's source code.

Reply:
What if you what have own test suite and you do use `WP_RUN_CORE_TESTS`? 
What if you running locally with `WP_DEBUG` set and want to debug issue with this code?

See https://github.com/WordPress/gutenberg/issues/45912. I think it is time to solve this problem.

Reply:
>What if you what have own test suite and you do use `WP_RUN_CORE_TESTS`?
`WP_DEBUG` should covers that use case.

>What if you running locally with `WP_DEBUG` set and want to debug issue with this code?
Do you mean if `WP_DEBUG` is set to `true`? 

Let's say there is a specific reason that a developer needs to disable the `WP_DEBUG` constant in order to debug or analyze the return of the static cache value. In this specific scenario, the developer can comment out the conditional check and do their debug/analyze when in development mode.  I think this is would be a rare or highly specific scenario.

@spacedmonkey are there scenarios where you see where this is _essential_ for developers?

Reply:
The code should not act different with debug and debug on. This is how it is currently "works" and it is weird and unexpected. 

Hardcoded `WP_RUN_CORE_TESTS` feels like a hacky solution, a filter feels a lot cleaner solution.

Reply:
>The code should not act different with debug and debug on. This is how it is currently "works" and it is weird and unexpected.

Why? In development or in test modes are different. There are other similar implementations using the constants throughout Core. This one of the purposes of the constants.

I respectfully disagree with adding a filter at this time per the reasons here https://github.com/WordPress/wordpress-develop/pull/3887#discussion_r1085425732. 

So @spacedmonkey, are you advocating to remove the `WP_DEBUG` and `WP_RUN_CORE_TESTS` constants completely?

Reply:
>Hardcoded `WP_RUN_CORE_TESTS` feels like a hacky solution, a filter feels a lot cleaner solution.

Why? That's the purpose of the constant, ie to know the automated test suite is running.

Reply:
What if you using WordPress unit tests in your own plugin and you not setting `WP_RUN_CORE_TESTS` and have a custom bootstrap for your own tests?

Code behaviour different in debug is bad for a number of reasons. But the main reason is, the code should act as much like production as possible. If the bug is in this code and does not when trying to debug, it is could be impossible to find the issue. You would have to run first with debug disable to see an error. This is confusing and make DUX harder.

Reply:
>What if you using WordPress unit tests in your own plugin and you not setting `WP_RUN_CORE_TESTS` and have a custom bootstrap for your own tests?

I wouldn't recommend using Core's test constant as that's for Core itself. Instead, use `WP_DEBUG`. One of the advantages of turning on `WP_DEBUG` is also to test for notices and deprecations in your test suite.

Reply:
I am just saying that, `WP_DEBUG` is a debug flag. It having a different side effect of disabling caching could have side effected we do not expect.  Some people mistakenly will enable debug in production. They may for example do this. 

```
define('WP_DEBUG', true);
define('WP_DEBUG_LOG', true);
define('WP_DEBUG_DISPLAY', false);
```

Where errors are logged by not displayed. Doing this would then tank performance of your site as it disables this cache.

Reply:
> A filter also changes how production works. It accomplishes the exact same thing as the constants but comes with disadvantages for performance, BC, and publicly exposing for production too.

Yes, but you are the developer can easily change it in your own code. With a custom bootstrap or a filter in a mu-plugin. That is the point.

Reply:
Which implementation provides the most benefits? Let's explore.

**Option 1: Using the filter**

Pros:
* Independent of the constants
* Can have `WP_DEBUG` without hooking into the filter providing testers / support / devs the flexibility to test in prod without using `WP_DEBUG`

Cons:
* Performance impact: the filter must be invoked each time the function runs. https://github.com/WordPress/wordpress-develop/pull/3887#issuecomment-1402055695, it would run on each of the "40+ calls" decreasing the performance savings.
* Could make debugging harder if a script, theme, or plugin hooks into the filter to ignore the static cache.
* The filter must be BC maintained.
* Unknown if exposing a filter is necessary for production.

**Option 2: Using constants**

Pros:
* Can aid in debugging as the constant's state does not change during a request/cycle.
* Consistent in Core.
* No extra code is needed in the Core's test suites.
* More performant than a filter.
* No BC concerns.
* Easily changed in the future.

Cons:
* Could be problematic for custom plugin/theme test suites in the following scenario: a test suite where `WP_DEBUG` is `false` and their code is using this function within the call stack. (Unknown if this scenario exists)

Reply:
I do appreciate what you're saying @spacedmonkey. I do see your points. But right now, it's unknown if there are plugin/theme test suites that are not turning on `WP_DEBUG` or want the means to separate `WP_DEBUG` from "in development" mode.

There are good reasons for when to use a filter vs constants. But in this specific function at this time, IMO the constants provide more benefits with less risks.

That said, if there are reported issues from extenders, then a filter or some other implementation can be explored.

Reply:
> then a filter or some other implementation can be explored.

Hopefully not before there is a (well defined and easy to turn on/off) "development mode" :)